File: CSPrograms/Usernames/Username.cs
using System;
using System.Text.RegularExpressions;

public class Username
{
    public static bool Validate(string username)
    {
        var Invalid = false;
        var Valid = true;

        if (username.Length < 6 || username.Length > 16)
            return Invalid;

        if (!isAlphaNumericWithHyphen(username) || !HasOneHyphen(username))
            return Invalid;

        if (!StartsWithLetter(username) || EndsWithHyphen(username))
            return Invalid;

        return Valid;
    }
    public static bool EndsWithHyphen(string username)
    {
        return username.EndsWith("-");
    }

    public static bool StartsWithLetter(string username)
    {
        return (username[0] >= 'A' && username[0] <= 'Z') ||
               (username[0] >= 'a' && username[0] <= 'z');
    }
    public static bool HasOneHyphen(string username)
    {
        return username.IndexOf('-') == username.LastIndexOf('-');
    }
    public static Boolean isAlphaNumericWithHyphen(string str)
    {
        Regex rg = new Regex(@"^[A-Za-z\d-]+$");
        return rg.IsMatch(str);
    }
}

File: JSPrograms/Validations/Username.js
(function UserNameValidator () {
    IsCorrectLength = function (str) {
        return str.length >= 6 && str.length <= 16
    }

    IsAlphaNumericWithHyphens = function (str) {
        return str.match(/^[a-z0-9-]+$/i);
    }

    HasOnlyOneHyphen = function (str) {
        return str.indexOf("-") === str.lastIndexOf("-")
    }

    StartsWithLetter = function (str) {
        return str.match(/^[a-z]/i);
    }

    EndsWithHyphen = function (str) {
        return str.match(/[-]$/i);
    }

    validate = function (username) {
        var Invalid = "Invalid Username."
        var Valid = "Valid Username.";

        if (IsCorrectLength(username) &&
            IsAlphaNumericWithHyphens(username) &&
            HasOnlyOneHyphen(username) &&
            StartsWithLetter(username) &&
            !EndsWithHyphen(username))
            return Valid;
        else return Invalid;
    }
})();